flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/benchmark/RescalingBenchmarkTest.java

```diff
@@ -0,0 +1,144 @@
+import org.apache.flink.runtime.state.storage.FileSystemCheckpointStorage;
+import org.apache.flink.streaming.api.functions.KeyedProcessFunction;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.util.Collector;
+import org.apache.flink.util.TestLogger;
+
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+import java.util.Iterator;
+import java.util.Random;
+
+/** Test Rescaling benchmark. */
+public class RescalingBenchmarkTest extends TestLogger {
+
+    private static final Random random = new Random(0);
+    @ClassRule public static TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+    @Test
+    public void testScalingOut() throws Exception {
+
+        RescalingBenchmark benchmark =
+                new RescalingBenchmarkBuilder()
+                        .setMaxParallelism(128)
+                        .setParallelismBefore(1)
+                        .setParallelismAfter(2)
+                        .setManagedMemorySize(512 * 1024 * 1024)
+                        .setCheckpointStorageAccess(
+                                new FileSystemCheckpointStorage(temporaryFolder.newFolder().toURI())
+                                        .createCheckpointStorage(new JobID()))
+                        .setStateBackend(new EmbeddedRocksDBStateBackend(true))
+                        .setStreamRecordGenerator(new IntegerRecordGenerator())
+                        .setStateProcessFunctionSupplier(() -> new TestKeyedFunction())
+                        .build();
+        benchmark.setUp();
+        benchmark.prepareStateAndHarnessForSubtask(0);
+        benchmark.rescale();
+        benchmark.closeHarnessForSubtask();
+        benchmark.tearDown();
+    }
+
+    @Test
+    public void testScalingIn() throws Exception {
+        RescalingBenchmark benchmark =
+                new RescalingBenchmarkBuilder()
+                        .setMaxParallelism(128)
+                        .setParallelismBefore(2)
+                        .setParallelismAfter(1)
+                        .setManagedMemorySize(512 * 1024 * 1024)
+                        .setCheckpointStorageAccess(
+                                new FileSystemCheckpointStorage(temporaryFolder.newFolder().toURI())
+                                        .createCheckpointStorage(new JobID()))
+                        .setStateBackend(new EmbeddedRocksDBStateBackend(true))
+                        .setStreamRecordGenerator(new IntegerRecordGenerator())
+                        .setStateProcessFunctionSupplier(() -> new TestKeyedFunction())
+                        .build();
+        benchmark.setUp();
+        benchmark.prepareStateAndHarnessForSubtask(0);
+        benchmark.rescale();
+        benchmark.closeHarnessForSubtask();
+        benchmark.tearDown();
+    }
+
+    @NotThreadSafe
```

Comment:
Since we only generate and process the record one by one, why we have to add such annotation here?